src/main.rs

```diff
@@ -1,7 +1,160 @@
-use crossterm::event::{Event, KeyCode, KeyEvent, KeyModifiers};
-use crossterm::{event, terminal};
-use std::time::Duration;
 use anyhow::Result;
+use crossterm::cursor::{Hide, MoveTo, Show};
+use crossterm::event::{self, Event, KeyEvent};
+use crossterm::style::Print;
+use crossterm::terminal::{self, Clear, ClearType::*};
+use crossterm::queue;
+use std::io::{self, BufWriter, Stdout, Write};
+
+pub struct Kilo {
+    cursor: Cursor,
+    dimentions: Dimentions,
+    is_running: bool,
+    stdout: BufWriter<Stdout>,
+}
+
+impl Kilo {
+    pub fn new() -> Result<Self> {
+        Ok(Self {
+            cursor: Cursor::new(0, 0),
+            dimentions: Dimentions::from(terminal::size()?),
+            is_running: true,
+            stdout: BufWriter::new(io::stdout()),
+        })
+    }
+
+    pub fn run(&mut self) -> Result<()> {
+        let _override = RawModeOverride::new()?;
+
+        while self.is_running {
+            self.render()?;
+            self.process_input()?;
+        }
+
+        Ok(())
+    }
+
+    fn terminate(&mut self) -> Result<()> {
+        self.is_running = false;
+        queue!(self.stdout, Clear(All), MoveTo(0, 0))?;
+        Ok(())
+    }
+
+    fn render(&mut self) -> Result<()> {
+        queue!(self.stdout, Hide)?;
+        queue!(self.stdout, MoveTo(0, 0))?;
+
+        self.render_rows()?;
+
+        queue!(self.stdout, Show)?;
+        queue!(self.stdout, MoveTo(self.cursor.x, self.cursor.y))?;
+
+        self.stdout.flush()?;
+        Ok(())
+    }
+
+    fn render_rows(&mut self) -> Result<()> {
+        for row in 0..self.dimentions.height {
+            if row == self.dimentions.height / 3 {
+                let message_row = self.make_message_row();
+                queue!(self.stdout, Print(message_row))?;
+            } else {
+                queue!(self.stdout, Print("~"))?;
+            }
+
+            queue!(self.stdout, Clear(UntilNewLine))?;
+            if row < self.dimentions.height - 1 {
```

Comment:
The 0/1 indexing conversion should ideally be the responsibility of the `Dimensions` class, consider making access to these members private and provide adjusted public getters.

Reply:
Sounds reasonable. I'm still hesitant on restricting access, but helper methods sound ok.